tuf-spec.md

```diff
@@ -1116,16 +1116,29 @@ repo](https://github.com/theupdateframework/specification/issues).
   cycle, report the potential freeze attack.  On the next update cycle, begin
   at step 0 and version N of the root metadata file.
 
-  * **1.9**. **If the timestamp and / or snapshot keys have been rotated, then
-  delete the trusted timestamp and snapshot metadata files.** This is done in
-  order to recover from fast-forward attacks after the repository has been
-  compromised and recovered. A _fast-forward attack_ happens when attackers
-  arbitrarily increase the version numbers of: (1) the timestamp metadata, (2)
-  the snapshot metadata, and / or (3) the targets, or a delegated targets,
-  metadata file in the snapshot metadata. Please see [the Mercury
+  * **1.9**. **Fast-forward attack recovery** A _fast-forward attack_ happens
+  when attackers arbitrarily increase the version numbers in any of the
+  timestamp, snapshot, targets, or delegated targets metadata. To recover from
+  fast-forward attacks after the repository has been compromised and recovered,
+  certain metadata files need to be deleted as specified in this section.
+  Please see [the Mercury
   paper](https://ssl.engineering.nyu.edu/papers/kuppusamy-mercury-usenix-2017.pdf)
   for more details.
 
+    * **1.9.1**. **Targets recovery** If a threshold of targets keys have been
+    removed in the new trusted root metadata compared to the previous trusted
+    root metadata, delete the old top-level targets and snapshot metadata
+    files.
```

Comment:
Why do we delete the snapshot metadata in this case?   If that key was compromised, won't it be rotated as well independently?

Reply:
What do you mean by "it will be rotated independently"?

If an attacker, who controls the top-level targets key, increases the version number in targets to  MAX_INT, the benign snapshot role might unwittingly adopt MAX_INT as the latest version number for targets. So we need to also delete snapshot after the compromised targets key has been rotated, otherwise we wouldn't get past verification step 3.3.3.

Reply:
I would think you could just rotate snapshot in this case too.  Or would this not work for some reason?

Reply:
You are right, the consequences in regards to ffwd-recovery would be the same (through 1.9.2.), but with the overhead of rotating an "uncompromised" key.

If we do that, we should require, or at least strongly recommend, to always rotate snapshot, if targets is compromised.  And we should also generalize this for all roles. That is, we should specify:
- If top-level targets is rotated, also rotate snapshot (as repo) and delete trusted top-level targets (as client)
- If snapshot is rotated, also rotate timestamp (as repo) and delete trusted snapshot (as client) 
- If timestamp is rotated, delete trusted timestamp (as client)

These keys are all rotated via root, so the overhead seems feasible. But for consistency, we should do the same for delegated targets, i.e. we'd need to specify:
- If a delegated targets key is rotated, also rotate snapshot (as repo) and delete trusted delegated targets (as client)

Now this would mean that rotation of delegated targets keys in a delegating targets file, would also require rotating the snapshot keys in the root file, which doesn't seem desirable.

Reply:
I'm not sure why all of these rotations are needed.  Isn't this only true the case of a fast-forward attack?  Why in general do you need to rotate the snapshot when there hasn't been a FF attack on the targets role?

Also, if there is a FF attack on a delegated key, I think you don't need to rotate snapshot.  The key that was used to sign will not be trusted anymore (post "rotation") so delegations will end up pointing at a new delegated target.  

I'd also like to say we're not talking about rotation in a TAP 8 sense here.  TAP 8 has another mechanism for dealing with these sorts of issues.

Reply:
> I'm not sure why all of these rotations are needed. Isn't this only true the case of a fast-forward attack? Why in general do you need to rotate the snapshot when there hasn't been a FF attack on the targets role?

Yes, they are indeed only needed in case there was an FFWD attack, so we could add that extra bit to the recommendation. But does the repo know for sure that there was or wasn't an ffwd-attack? To me it seems safer to just assume it, so that clients can always resume updates. 

> Also, if there is a FF attack on a delegated key, I think you don't need to rotate snapshot. The key that was used to sign will not be trusted anymore (post "rotation") so delegations will end up pointing at a new delegated target.

Let's say an attacker compromises the delegated target `django` and bumps its version number in 'django.json' from `n` to `max_int`. The `snapshot` role  learns that 'django.json' is at `max_int` and, with good intentions,  writes `'django.json' : {'version': max_int}` to 'snapshot.json'.  Now the repo maintainer realize that  `django`'s key was compromised, replace it via its delegating role, and reset the version number to `n+1` in 'django.json', which `snapshot` adopts in 'snapshot.json'. Now since the `snapshot` key was not replaced, because it was not compromised, the client will not remove the trusted 'snapshot.json' that lists `max_int`, and thus will refuse to update to the lower `n+1` listed in the new 'snapshot.json' due to the rollback check in [3.3.3](https://github.com/theupdateframework/specification/blame/0f56aee998db130e6e1c43c5c7141217436b6d33/tuf-spec.md#L1188-L1195) (before this PR) or [4.5.2.3](https://github.com/theupdateframework/specification/blame/132f96a5c0d56f8b09c3bb775fa0c848298b2ba1/tuf-spec.md#L1274-L1278) (after this PR).

Or is it safe to assume that an attacker can't trick another non-compromised role into signing a wrong version number?


> I'd also like to say we're not talking about rotation in a TAP 8 sense here. TAP 8 has another mechanism for dealing with these sorts of issues.

Correct. By "rotation" above, I mean "a threshold of keys is removed". This PR also uses the latter wording instead of the former, in order to clarify this.

Reply:
Okay, after talking with @lukpueh , I think we're on the same page.  I was misremembering that something like the keyid is listed in the file name for a delegated targets role.  We do agree that cases where you know a fast forward attack did not occur (such as situations where a repository compromise did not happen) do not need to have the snapshot, etc. metadata deleted.